FreeRTOS_IP.c

```diff
@@ -2323,22 +2325,17 @@ static eFrameProcessingResult_t prvProcessIPPacket( IPPacket_t * pxIPPacket,
             pxIPHeader->usFragmentOffset = 0U;
         #endif
 
-        /* Update the checksum because the ucTypeOfMessage member in the header
-         * has been changed to ipICMP_ECHO_REPLY.  This is faster than calling
-         * usGenerateChecksum(). */
-
-        /* due to compiler warning "integer operation result is out of range" */
-
-        usRequest = ( uint16_t ) ( ( uint16_t ) ipICMP_ECHO_REQUEST << 8 );
+        #if ( ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM == 0 )
+            {
+                /* calculate the IP header checksum, in case the driver won't do that. */
+                pxIPHeader->usHeaderChecksum = 0x00U;
+                pxIPHeader->usHeaderChecksum = usGenerateChecksum( 0U, ( uint8_t * ) &( pxIPHeader->ucVersionHeaderLength ), ipSIZE_OF_IPv4_HEADER );
+                pxIPHeader->usHeaderChecksum = ~FreeRTOS_htons( pxIPHeader->usHeaderChecksum );
 
-        if( pxICMPHeader->usChecksum >= FreeRTOS_htons( 0xFFFFU - usRequest ) )
-        {
-            pxICMPHeader->usChecksum = pxICMPHeader->usChecksum + FreeRTOS_htons( usRequest + 1U );
-        }
-        else
-        {
-            pxICMPHeader->usChecksum = pxICMPHeader->usChecksum + FreeRTOS_htons( usRequest );
-        }
+                /* calculate the ICMP checksum for an outgoing packet. */
+                ( void ) usGenerateProtocolChecksum( ( uint8_t * ) pxICMPPacket, pxNetworkBuffer->xDataLength, pdTRUE );
+            }
+        #endif /* if ( ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM == 0 ) */
```

Comment:
It would be desirable to clear the ICMP checksum 
~~~c
#else
{
    /* Many EMAC peripherals will only calculate the ICMP checksum
     * correctly if the field is nulled beforehand. */
    pxICMPHeader->usChecksum = 0U;
}
~~~

Reply:
But the IP header checksum is different?
I don't see the TCP checksum nulled either?

Reply:
We should ask the chip designers why the ICMP checksum must be zero, whereas the IP- and other protocol-checksums may have a random value. Strange indeed.

Reply:
okay, so I will zero out the ICMP header checksum, but leave the IP header checsum unchanged.
right?

Reply:
> Update: the above avoids code like this

Oh, I see.  Should I remove this code in this PR ?

Reply:
Talking about this code in NetworkInterface.c:
~~~c
if( pxPacket->xICMPPacket.xIPHeader.ucProtocol == ( uint8_t ) ipPROTOCOL_ICMP )
{
    pxPacket->xICMPPacket.xICMPHeader.usChecksum = ( uint16_t ) 0u;
}
~~~
As I am not sure if users always have the same version of FreeRTOS_IP.c and NetworkInterface.c, it is maybe safer to keep it for now.